In [8]:
from keras.models import Sequential
from keras.layers import Dense
import numpy as np
import pandas as pd


# Seeding to make sure that we get exact same result everytime the code is run

In [64]:
seed = 7
np.random.seed(seed)

# Predictive Analysis problem. Predict the chance that a person will buy a bike

In [40]:
input_file = 'KerasDownloads/BBCN.csv'
input_df = pd.read_csv(input_file)

# Preview

In [41]:
print input_df.shape
input_df.head()

(500, 12)


MaritalStatus  Gender  YearlyIncome  TotalChildren  NumberChildrenAtHome  \
0              5       1           9.0              2                     0   
1              5       1           6.0              3                     3   
2              5       1           6.0              3                     3   
3              5       2           7.0              0                     0   
4              5       2           8.0              5                     5   

   EnglishEducation  HouseOwnerFlag  NumberCarsOwned  CommuteDistance  Region  \
0                 5               1                0                2       2   
1                 5               0                1                1       2   
2                 5               1                1                5       2   
3                 5               0                1               10       2   
4                 5               1                4                2       2   

   Age  BikeBuyer  
0    5          1  
1    4          1  
2    4          1  
3    5          1  
4    5          1

# Split the data to input and target variable

In [51]:
array = input_df.values
X = array[:, 0:11]
Y = array[:, 11]

# Build the model

Models in keras are defined as the sequence of layers

  First layer is the input layer and the subsequential are hidden layers except the output layer. This forms a `Sequential` model.
  
  Neurons in the layer are fully connected, meaning every neuron in ith layer is connected to every neuron in (i+1)th layer. For this we use `Dense` layering structure
  
  `input_dim` is the number of features used to predict the output
  
  

In [58]:
model = Sequential()
model.add(Dense(11, input_dim = 11, kernel_initializer = 'uniform', activation = 'relu')) # input layer
model.add(Dense(8, kernel_initializer = 'uniform', activation = 'relu')) # 1st hidden layer
model.add(Dense(8, kernel_initializer = 'uniform', activation = 'relu')) # 2nd hidden layer
model.add(Dense(8, kernel_initializer = 'uniform', activation = 'relu')) # 3nd hidden layer
model.add(Dense(1, kernel_initializer = 'uniform', activation = 'sigmoid')) # output layer


 # Model compilation
 
 This uses the backend Tensorflow here.
 
 `metrics` are used to judge the performance of our model. Here we are using `accuracy` which is good only for symmetric data sets, where the class distribution is 50/50.In our case, classes are balanced.
 
 Use `crossentropy` cost function in case of classification problems.

In [59]:
model.compile(loss = 'binary_crossentropy',optimizer = 'adam', metrics = ['accuracy'])

# Model fitting

`epochs` = number of times to run the model over the training data

`batch_size` = number of samples that going to be propagated through the network

See this: https://stats.stackexchange.com/questions/153531/what-is-batch-size-in-neural-network


In [62]:
model.fit(X, Y, epochs = 200, batch_size = 10)

Epoch 1/200
500/500 [==============================] - 0s - loss: 0.6525 - acc: 0.6260     
Epoch 2/200
500/500 [==============================] - 0s - loss: 0.6502 - acc: 0.6340     
Epoch 3/200
500/500 [==============================] - 0s - loss: 0.6479 - acc: 0.6280     
Epoch 4/200
500/500 [==============================] - 0s - loss: 0.6447 - acc: 0.6260     
Epoch 5/200
500/500 [==============================] - 0s - loss: 0.6421 - acc: 0.6420     
Epoch 6/200
500/500 [==============================] - 0s - loss: 0.6408 - acc: 0.6280     
Epoch 7/200
500/500 [==============================] - 0s - loss: 0.6399 - acc: 0.6320     
Epoch 8/200
500/500 [==============================] - 0s - loss: 0.6386 - acc: 0.6380     
Epoch 9/200
500/500 [==============================] - 0s - loss: 0.6350 - acc: 0.6380     
Epoch 10/200
500/500 [==============================] - 0s - loss: 0.6337 - acc: 0.6420     
Epoch 11/200
500/500 [==============================] - 0s - loss: 0.6306 - acc

500/500 [==============================] - 0s - loss: 0.5849 - acc: 0.6780     
Epoch 89/200
500/500 [==============================] - 0s - loss: 0.5845 - acc: 0.6860     
Epoch 90/200
500/500 [==============================] - 0s - loss: 0.5841 - acc: 0.6800     - ETA: 0s - loss: 0.5822 - acc: 0.68
Epoch 91/200
500/500 [==============================] - 0s - loss: 0.5802 - acc: 0.6860     
Epoch 92/200
500/500 [==============================] - 0s - loss: 0.5815 - acc: 0.6820     
Epoch 93/200
500/500 [==============================] - 0s - loss: 0.5796 - acc: 0.6840     
Epoch 94/200
500/500 [==============================] - 0s - loss: 0.5780 - acc: 0.6920     
Epoch 95/200
500/500 [==============================] - 0s - loss: 0.5778 - acc: 0.6900     
Epoch 96/200
500/500 [==============================] - 0s - loss: 0.5837 - acc: 0.6740     
Epoch 97/200
500/500 [==============================] - 0s - loss: 0.5777 - acc: 0.6880     
Epoch 98/200
500/500 [=========================

500/500 [==============================] - 0s - loss: 0.5522 - acc: 0.7040     
Epoch 175/200
500/500 [==============================] - ETA: 0s - loss: 0.5458 - acc: 0.700 - 0s - loss: 0.5542 - acc: 0.6940     
Epoch 176/200
500/500 [==============================] - 0s - loss: 0.5551 - acc: 0.6980     
Epoch 177/200
500/500 [==============================] - 0s - loss: 0.5465 - acc: 0.7000     
Epoch 178/200
500/500 [==============================] - 0s - loss: 0.5542 - acc: 0.6960     
Epoch 179/200
500/500 [==============================] - 0s - loss: 0.5475 - acc: 0.7060     
Epoch 180/200
500/500 [==============================] - 0s - loss: 0.5587 - acc: 0.7020     
Epoch 181/200
500/500 [==============================] - 0s - loss: 0.5472 - acc: 0.6960     
Epoch 182/200
500/500 [==============================] - 0s - loss: 0.5503 - acc: 0.7020     
Epoch 183/200
500/500 [==============================] - 0s - loss: 0.5562 - acc: 0.7000     
Epoch 184/200
500/500 [=============

# Model score

In [63]:
scores = model.evaluate(X, Y)
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

 32/500 [>.............................] - ETA: 0sacc: 71.80%


The above model is not much better than flipping a coin, which is real bad. so we'll tune the network

# One thing we could do is normalize the data and repeat the process.

  See that we'll be changing the input file.

  We see that accuracy changed from 50.60% to 62.00%

# Add another hidden layer

   We did a better job by adding a new layer. Accuracy change from 62.00% to 63.40%
   
# Change number of neurons in a hidden layer


# Increase the number of `epochs`

   This helped in bringing the accuracy to 71.80%
